Correlations: Add transformations to Explore Editor

gelicia opened the pull request:
**What is this feature?**

Adds the ability to create one or more correlation transformations from the Explore correlation editor. 

**Why do we need this feature?**

Transformations are almost necessary for creating correlations on logs. Logs' data format generally has more than one piece of information per field, notably the log line field and the labels field. Transformations were created to parse out information people may want, via a regular expression or logfmt. 

**Who is this feature for?**

People who want to create correlations where data may be within in one field amongst other data.

**Which issue(s) does this PR fix?**:

Fixes #75832 

**Special notes for your reviewer:**
Testing this with devenv, using Loki is probably your best bet. Logfmt will parse the devenv line, and labels can be parsed out with the following regex expressions: 
- `namespace":"([a-z]*)` - this is an unnamed capture group
- `namespace":"(?<test2>[a-z]*)` this is a named capture group

In the [ephemeral](https://ephemeral1511182175930gelicia.grafana-dev.net/), loki data is available in the `grafanacloud-ephemeral1511182175930gelicia-usage-insights` Loki datasource. Doing a logfmt transformation against a Line field adds a lot of variables. Adding a regex transformation on labels works with the following 
- `instance_type":"([a-z]*)` - this is an unnamed capture group
- `instance_type":"(?<name>[a-z]*)` - this is a named capture group

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

gelicia committed:
Add transformation add modal and use it

gelicia committed:
Hook up saving

gelicia committed:
Add transformation vars to var list, show added transformations

gelicia commented:
/deploy-to-hg

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Form validation

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Remove type assertion, start building out transformation data in helper (WIP)

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Style expression better, add delete logic

gelicia committed:
Add ability to edit, additional styling on transformation card in helper

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
simplify styling, conditionally run edit set up logic

gelicia commented:
/deploy-to-hg

gelicia commented:
@diegoadams This is ready for you now! The see the PR description for a link to the ephemeral and some tips on what to try.

gelicia committed:
Keep more field information in function, integrate it with new editor

gelicia commented:
/deploy-to-hg

diegoadams commented:
@gelicia, in the editor mode, 

After going through the flow again and once I was in editor mode and clicked on a datalink and the view split, I wasn't sure what to do next, but once my eyes saw the Correlation details container I understood. 

I wonder if putting more emphasis on the correlation details container would make a better connection to what's needed next. I found the current lighter container too subtle for me to notice where to look. 

A bold 5px border in our primary blue could help make a clearer connection to the editor and highlight the correlations view.

<img width="1309" alt="corblue" src="https://github.com/grafana/grafana/assets/95694434/8dbc63f4-cb7c-4c56-908c-469e88273d2c">
 
we can also look at this for Phase 2

gelicia commented:
@diegoadams 
> Regarding transformations, do we have data on the average number users typically add? Also is there a max amount of transformations that could be added?
We dont - we save when people add a correlation but not any details about it. That said, the editor and just general use of transformations would limit people in a way we are trying to overcome. I would guess most people would have 1-2 transformations if they needed them, and 99% would have under 5 transformations in a single correlation.

I can add a confirmation modal for deleting. Id like other UI updates to the helper to go in its own ticket. Adding a border isn't a big deal but its much easier to iterate on these kinds of things in isolation. I created https://github.com/grafana/grafana/issues/76263

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Show default label on collapsed section, use deleteButton for confirmation of deleting transformations

gelicia committed:
Change transformation add calculations from function to hook, add label to collapsed header, add transformation tooltip

gelicia commented:
> I found a bit confusing that the label was generated automatically for me without me knowing

Fixed by adding the default value to the field, and showing the value of the field when the collapsible is collapsed. The styling is to somewhat mimic what we have for query row editors

> a one liner explaining what you can do with transformations in the collapse would also go a long way

Fixed with a tooltip in the collapsible title. I did this way instead of under the collapse so it wouldnt take up more vertical space.

> the lack of actionable feedback to make the form valid, may be a bit frustrating

This is a more complicated one that I would like to iterate over later. 

> changing the transformation type doesn't re-trigger the validation, so the button stays disabled until i touch one of the inputs even though the configuration is valid

Fixed.

gelicia commented:
@diegoadams @ifrost 

I added in a confirmation button when you delete a transformation. I'm using the built in `DeleteButton` component that is part of Grafana UI https://developers.grafana.com/ui/latest/index.html?path=/story/buttons-confirmbutton--delete

@ifrost brought up that whatever we do here, we should do the same thing to the existing transformation editor in Administration > Correlations to make it consistent. This is the result of a quick swapping out the buttons

https://github.com/grafana/grafana/assets/1533128/98199181-ddc3-448b-8506-af0358b3f035

As you can see, this does not look great, and the direction the buttons show up is not currently configurable. I recommend we acknowledge the inconsistency, leave the admin editor as it is for now and make changing this part of a separate ticket.

The explore editor looks better for this, fwiw:

https://github.com/grafana/grafana/assets/1533128/eb18aaf2-4f02-4ba0-93d7-9d0a0d061cd7

gelicia commented:
/deploy-to-hg

gelicia commented:
/deploy-to-hg

gelicia commented:
/deploy-to-hg

diegoadams commented:
> @diegoadams @ifrost
> 
> I added in a confirmation button when you delete a transformation. I'm using the built in `DeleteButton` component that is part of Grafana UI https://developers.grafana.com/ui/latest/index.html?path=/story/buttons-confirmbutton--delete
> 
> @ifrost brought up that whatever we do here, we should do the same thing to the existing transformation editor in Administration > Correlations to make it consistent. This is the result of a quick swapping out the buttons
> 
>  correditoradmin.mov 
> As you can see, this does not look great, and the direction the buttons show up is not currently configurable. I recommend we acknowledge the inconsistency, leave the admin editor as it is for now and make changing this part of a separate ticket.
> 
> The explore editor looks better for this, fwiw:
> 
>  correditorexplore.mov

Hi @gelicia, thanks for the visuals. I agree, we can address this at a later stage and leave it as is for now. Do you want me to add it to this ticket I created [76263](https://github.com/grafana/grafana/issues/76263 ) along with the border suggestion?

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia commented:
/deploy-to-hg

gelicia commented:
Bug: changing target datasource in dirty state doesnt persist variables

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Make correlation and editor dirty state distinctive and integrate, WIP

gelicia committed:
Track action pane for more detailed messaging/actions

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Better cancel modal logic

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia commented:
I added some better logic with modal messaging/actions when closing a pane or changing a datasource. It now tracks if the right query was changed or correlation was changed (dirty state) separately so the message can provide better details on what the consequences of the action are. 

<details>
  <summary>Modal logic</summary>
  
1. Close left pane - correlations links show in right pane, helper goes away
     1. Left query - is destroyed. No alert - user knows action’s consequences
     1. Left correlation links - are destroyed. No alert - user knows action’s consequences
     1. Right query dirty - alert “Closing this pane will cause the query in the right pane to be re-ran and links added to the result data. Would you like to save your correlation before continuing?” 
     1. Right correlation dirty - alert “Closing the left pane will cause the correlation in progress to be lost.Would you like to save your correlation before continuing?” 
     1. Both dirty - use correlation alert message
1. Close right pane - left pane unaffected, helper goes away
     1. Left query - no change, no alert
     1. Left correlation links - no change, no alert 
     1. Right query dirty - alert “Closing this pane will lose the changed query. Would you like to save your correlation before continuing?”
     1. Right correlation dirty - alert “Closing this pane will cause the correlation in progress to be lost. Would you like to save your correlation before continuing?”
     1. Both dirty - use correlation dirty message
1. Change DS left - regenerate links, helper goes away because variables changed
     1. Left query - is destroyed. No alert - user should know consequences
     1. Left correlation links - are destroyed. No alert - user should know consequences
     1. Right query dirty - No alert (But this might need to change). Variables will change, meaning any used variables may be invalid. We have no way of knowing if the right query used variables or not, so they might have just changed an option without using variables and an alert here might cause more alarm than necessary.
     1. Right correlation dirty - alert “Changing the datasource will cause the correlation in progress to be lost. Would you like to save before continuing?” 
     1. Both right dirty - use correlation dirty message
1. Change DS right - helper stays, right query goes away
     1. Left query - No change, no alert
     1. Left correlation links - No change, no alert 
     1. Right query dirty - alert “Changing the datasource will lose the changed query. Would you like to save your correlation before continuing?”
     1. Right correlation dirty - no alert. This only changes the default label, which would not destroy a user action.
     1. Both dirty - use query dirty message

</details>

gelicia committed:
Remove changes to adminsitration transformation editor

gelicia commented:
/deploy-to-hg

gelicia commented:
@diegoadams 
> Hi @gelicia, thanks for the visuals. I agree, we can address this at a later stage and leave it as is for now. Do you want me to add it to this ticket [76263](https://github.com/grafana/grafana/issues/76263) along with the border suggestion?

Removed the changes from admin and the corresponding test.

gelicia committed:
Remove debugging line

ifrost commented on public/app/features/explore/CorrelationEditorModeBar.tsx:
```diff
@@ -9,44 +9,70 @@ import { Button, HorizontalGroup, Icon, Tooltip, useStyles2 } from '@grafana/ui'
+          }
+        }
+
+        // todo do
```
What needs to be done here?

gelicia replied:
Shouldnt be needed. Removed.

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Remove unneeded comment

harisrozajac reviewed (commented):
`Exit correlation editor` button doesn't work. Otherwise everything else looks great. I've played with it both in local and ephemeral instance and everything is working nicely.

gelicia commented:
> `Exit correlation editor` button doesn't work. Otherwise everything else looks great. I've played with it both in local and ephemeral instance and everything is working nicely.

Ah yep, spent too much time thinking about two exit flows and forgot a third one. Will fix!

gelicia committed:
Add in logic for closing editor mode

gelicia commented:
/deploy-to-hg

harisrozajac reviewed (approved):
🚀 great work!

gelicia committed:
Add tests for modal logic

gelicia commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+import { showModalMessage } from './correlationEditLogic';
+
+// note, closing the editor does not care if isLeft is true or not. Both are covered for regression purposes.
+describe('correlationEditLogic', function () {
```
@ifrost - added tests :)

ifrost commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${true}  | ${false} | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${false} | ${true}  | ${'Closing the pane will cause the query in the right pane to be re-ran and links added to that data. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${true}  | ${true}  | ${true}  | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${false} | ${false} | ${false} | ${undefined}
```
It may not happen with each data source but sometimes changing the data source without making any changes causes the modal to appear. Could it be because the editor calls `onChange`? 

https://github.com/grafana/grafana/assets/745532/1fd44ae0-9162-444f-937c-805300e1f015

gelicia replied:
Yeah :/ and I'm not quite sure what the solution is.

The order of events is this

1. User changes the datasource to cloudwatch. This completes. The onChange function that we use to track a dirty query is passed into the query editor
2. Cloudwatch's query editor internally runs a function that calls the onChange function passed in https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/cloudwatch/components/QueryEditor/QueryEditor.tsx#L28 - If you look at the queries before and after the change, it sets up the data structure for CW with log groups, region, an empty logGroups array, etc. 
3. The onChange function is executed and the query gets marked as dirty even without any user interaction interacting with the editor.

I need to investigate some more, but I don't know how to make the distinction between a user performed query change and an automatic one, when the editor itself doesn't make that distinction at all. I thought that ensuring the query was not marked as dirty after the DS changed would be enough, but the query editor for the plugin is mounted after and I don't think there's any sort of state change that happens after that mount occurs.

ifrost commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+    action                                                      | isLeft   | dirCor   | dirQuer  | expected
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${true}  | ${false} | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${false} | ${true}  | ${'Closing the pane will lose the changed query. Would you like to save before continuing?'}
```
The modal seems to be shown twice when closing the pane:


https://github.com/grafana/grafana/assets/745532/581dd28f-ed71-4ace-8d18-a886cff18aa8

gelicia replied:
Fixed!

ifrost commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${true}  | ${true}  | ${true}  | ${'Changing the datasource will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_EDITOR}      | ${false} | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_EDITOR}      | ${false} | ${true}  | ${false} | ${'Closing the editor will cause the correlation in progress to be lost. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_EDITOR}      | ${false} | ${false} | ${true}  | ${'Closing the editor will lose the changed query. Would you like to save before continuing?'}
```
In what sense "the query is lost"? I tried closing the right pane after changing the query and it seems to run exactly the same query without losing anything:


https://github.com/grafana/grafana/assets/745532/5ec7128e-51b4-4f53-8687-8d8fa8a2b4a2

gelicia replied:
Ah this is true. I might change this to say "Closing the editor will remove the variables, and your changed query may no longer be valid. Would you like to save before continuing?" in this instance

ifrost commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+  it.each`
+    action                                                      | isLeft   | dirCor   | dirQuer  | expected
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CLOSE_PANE}        | ${false} | ${true}  | ${false} | ${'Closing the pane will cause the correlation in progress to be lost. Would you like to save before continuing?'}
```
Changing the label doesn't mark a correlation as dirty. Is it expected?


https://github.com/grafana/grafana/assets/745532/796bb9a7-ddbb-4f14-8667-4dd7e01aef24

gelicia replied:
No - changing the label means the correlation dirty state is true. This was a regression when I used the default value as a default value instead of the default value being blank and only using the default value on save. 

And actually I think I'll be able to tackle this and the mixed datasource problem at the same time, so I'll be looking into this more.

ifrost reviewed (commented):
<details><summary>Transformations vars are not interpolated inside the editor</summary>

After adding transformations I was not able to use vars inside the editor / right pane. After saving everything works fine. It looks like only when we're inside the editor, then transformation vars are not used when interpolating.

https://github.com/grafana/grafana/assets/745532/44991509-c0aa-4ab6-a845-20cb1dd1d308

</details>

Also probably not related to this PR but worth checking:

<details><summary>Mixed data source is used inside the default label instead of the underlying data source</summary>

<img width="1526" alt="mixed-bug" src="https://github.com/grafana/grafana/assets/745532/ed5a0a63-96f4-4de4-bf67-075c81bc9556">

</details>

<details><summary>Regex is case insenstive</summary>

I think it's been like this since we added this but I wonder if it shouldn't base case sensitive to allow capturing more specific cases. Or at least at it to the tooltip / docs.

<img width="680" alt="case sensitive" src="https://github.com/grafana/grafana/assets/745532/dd96dba6-e7c7-4172-851e-47c81de8f577">


</details>

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Use state to build vars list for helper

gelicia commented:
> Transformations vars are not interpolated inside the editor

fixed!

> Mixed data source is used inside the default label instead of the underlying data source

fixed!

> Regex is case insenstive
 
I did this on purpose to increase the likelihood that people will have matches and I'd like to keep it that way for now. I added it to the tooltip.

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
WIP

gelicia committed:
Fix labels and dirty state

gelicia committed:
Fix bad message and stop exiting mode if discard action is performed

gelicia committed:
Fix tests

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

ifrost commented on public/app/features/correlations/utils.ts:
```diff
@@ -90,3 +91,19 @@ export const createCorrelation = async (
+  }
+};
+
+export const generateDefaultLabel = async (sourcePane: ExploreItemState, targetPane: ExploreItemState) => {
```
Ideally correlations utils should not rely on any Explore specific types. This function could get `datasourceRef` passed instead of entire `ExploreItemState`.

ifrost commented on public/app/features/explore/correlationEditLogic.test.ts:
```diff
@@ -0,0 +1,39 @@
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${false} | ${false} | ${true}  | ${'Changing the datasource will lose the changed query. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${false} | ${true}  | ${true}  | ${'Changing the datasource will lose the changed query. Would you like to save before continuing?'}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${true}  | ${false} | ${false} | ${undefined}
+    ${CORRELATION_EDITOR_POST_CONFIRM_ACTION.CHANGE_DATASOURCE} | ${true}  | ${true}  | ${false} | ${'Changing the datasource will cause the correlation in progress to be lost. Would you like to save before continuing?'}
```
Personally I find it confusing that I'm asked if I want to save the correlation at this point. I'd expect a simple question like "Your progress will be lost. Do you want to continue?". But I'll leave it up to UX.

ifrost reviewed (commented):
Is documentation being added as part of this PR?

gelicia commented:
> Is documentation being added as part of this PR?

Created https://github.com/grafana/grafana/issues/77457 and that will be my next task

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform

gelicia committed:
Update to not use unstable component and tweak default label

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into kristina/correlations-editor-transform
